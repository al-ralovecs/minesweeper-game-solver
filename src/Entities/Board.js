export default class Board {
  constructor(state) {
    this.state = state;

    this.tryFlagMines();
  }

  get height() {
    return this.state.length;
  }

  get width() {
    if (typeof this.boardWidth === 'undefined') {
      this.boardWidth = this.getBoardWidth();
    }

    return this.boardWidth;
  }

  getBoardWidth() {
    const width = this.state[0].length;

    if (this.height === 1) {
      return width;
    }

    for (let i = 1; i < this.height; i++) {
      if (width !== this.state[i].length) {
        throw Error();
      }
    }

    return width;
  }

  tryFlagMines() {
    let flags = [];

    for (let i = 0; i < this.height; i++) {
      let row = [];
      for (let j = 0; j < this.width; j++) {
        row.push(false);
      }
      flags.push(row);
    }

    for (let i = 0; i < this.height; i++) {
      for (let j = 0; j < this.width; j++) {
        let curNum = this.getCell(i, j);

        if (curNum < 1 || curNum !== this.countFreeSquaresAround(i, j)) {
          continue;
        }

        for (let ii = 0; ii < this.height; ii++) {
          for (let jj = 0; jj < this.width; jj++) {
            if (Math.abs(ii-i) <= 1 && Math.abs(jj-j) <= 1) {
              if (this.getCell(ii, jj) === -1 && ! flags[ii][jj]) {
                flags[ii][jj] = true;
              }
            }
          }
        }
      }
    }

    this.flags = flags;
  }

  getCell(y, x) {
    if (x >= 0 && x < this.width && y >= 0 && y < this.height) {
      return this.state[y][x];
    }

    return -10;
  }

  hasMine(y, x) {
    if (x >= 0 && x < this.width && y >= 0 && y < this.height) {
      return this.flags[y][x];
    }

    throw Error();
  }

  isBoundary(y, x) {
    if (this.getCell(y, x) !== -1) {
      return false;
    }

    const
      oU = y === 0,
      oD = y === (this.height - 1),
      oL = x === 0,
      oR = x === (this.width - 1);

    if (!oU && this.getCell(y-1, x) >= 0) {
      return true;
    }
    if (!oL && this.getCell(y, x-1) >= 0) {
      return true;
    }
    if (!oD && this.getCell(y+1, x) >= 0) {
      return true;
    }
    if (!oR && this.getCell(y, x+1) >= 0) {
      return true;
    }
    if (!oU && !oL && this.getCell(y-1, x-1) >= 0) {
      return true;
    }
    if (!oU && !oR && this.getCell(y-1, x+1) >= 0) {
      return true;
    }
    if (!oD && !oL && this.getCell(y+1, x-1) >= 0) {
      return true;
    }
    if (!oD && !oR && this.getCell(y+1, x+1) >= 0) {
      return true;
    }

    return false;
  }

  countFreeSquaresAround(y, x) {
    let result = 0;

    if (this.getCell(y-1, x) === -1) {
      result++;
    }
    if (this.getCell(y+1, x) === -1) {
      result++;
    }
    if (this.getCell(y, x-1) === -1) {
      result++;
    }
    if (this.getCell(y, x+1) === -1) {
      result++;
    }
    if (this.getCell(y-1, x-1) === -1) {
      result++;
    }
    if (this.getCell(y-1, x+1) === -1) {
      result++;
    }
    if (this.getCell(y+1, x-1) === -1) {
      result++;
    }
    if (this.getCell(y+1, x+1) === -1) {
      result++;
    }

    return result;
  }
}
