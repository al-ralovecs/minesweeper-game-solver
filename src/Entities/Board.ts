export default class Board {
    private readonly state: number[][];
    private flags: boolean[][];

    private height: number;
    private width: number;

    public constructor(state: number[][])
    {
        if (! Board.validate(state)) {
            throw Error();
        }

        this.state = state;

        this.flagMines();
    }

    public get getHeight(): number
    {
        if (typeof this.height === 'undefined') {
            this.height = this.state.length;
        }

        return this.height;
    }

    public get getWidth(): number
    {
        if (typeof this.width === 'undefined') {
            this.width = this.state[0].length;
        }

        return this.width;
    }

    public getValue(y: number, x: number): number
    {
        if ( y >= 0 && y < this.getHeight && x >= 0 && x < this.getWidth) {
            return this.state[y][x];
        }

        return -10;
    }

    public getSurroundingFreeSquaresCount(y: number, x: number): number
    {
        let result = 0;

        if (this.getValue(y-1, x) === -1) {
            result++;
        }
        if (this.getValue(y+1, x) === -1) {
            result++;
        }
        if (this.getValue(y, x-1) === -1) {
            result++;
        }
        if (this.getValue(y, x+1) === -1) {
            result++;
        }
        if (this.getValue(y-1, x-1) === -1) {
            result++;
        }
        if (this.getValue(y-1, x+1) === -1) {
            result++;
        }
        if (this.getValue(y+1, x-1) === -1) {
            result++;
        }
        if (this.getValue(y+1, x+1) === -1) {
            result++;
        }

        return result;
    }

    public isBoundary(y: number, x: number): boolean
    {
        if (this.getValue(y, x) !== -1) {
            return false;
        }

        const
            oU: boolean = y === 0,
            oD: boolean = y === (this.getHeight - 1),
            oL: boolean = x === 0,
            oR: boolean = x === (this.getWidth - 1);

        if (!oU && this.getValue(y - 1, x) >= 0) {
            return true;
        }
        if (!oL && this.getValue(y, x - 1) >= 0) {
            return true;
        }
        if (!oD && this.getValue(y + 1, x) >= 0) {
            return true;
        }
        if (!oR && this.getValue(y, x + 1) >= 0) {
            return true;
        }
        if (!oU && !oL && this.getValue(y - 1, x - 1) >= 0) {
            return true;
        }
        if (!oU && !oR && this.getValue(y - 1, x + 1) >= 0) {
            return true;
        }
        if (!oD && !oL && this.getValue(y + 1, x - 1) >= 0) {
            return true;
        }

        return !oD && !oR && this.getValue(y + 1, x + 1) >= 0;
    }

    public hasMine(y: number, x: number): boolean
    {
        if (y >= 0 && y < this.getHeight && x >= 0 && x < this.getWidth) {
            return this.flags[y][x];
        }

        throw Error();
    }

    private flagMines(): void
    {
        let flags: boolean[][] = [];

        for (let i = 0; i < this.getHeight; i++) {
            let row: boolean[] = [];
            for (let j = 0; j < this.getWidth; j++) {
                row.push(false);
            }
            flags.push(row);
        }

        for (let i: number = 0; i < this.height; i++) {
            for (let j: number = 0; j < this.width; j++) {
                let currentValue: number = this.getValue(i, j);

                if (currentValue < 1 || currentValue !== this.getSurroundingFreeSquaresCount(i, j)) {
                    continue;
                }

                for (let ii: number = 0; ii < this.getHeight; ii++) {
                    for (let jj:number = 0; jj < this.getWidth; jj++) {
                        if (Math.abs(ii-i) <= 1 && Math.abs(jj-j) <= 1) {
                            if (this.getValue(ii, jj) === -1 && ! flags[ii][jj]) {
                                flags[ii][jj] = true;
                            }
                        }
                    }
                }
            }
        }

        this.flags = flags;
    }

    private static validate(state: number[][]): boolean
    {
        let width: number = 0;
        for (let i = 0; i < state.length; i++) {
            if (0 === width) {
                width = state[i].length;
                continue;
            }

            if (state[i].length === width) {
                continue;
            }

            return false;
        }

        return true;
    }
}
