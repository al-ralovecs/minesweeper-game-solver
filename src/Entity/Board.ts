import Coordinate from "./Coordinate";
import countFlags from '../Routine/board.matrix.countTotalFlags';
import countFlagsAround from '../Routine/board.item.countFlagsAround';
import flagMines from '../Routine/board.matrix.flagMines';
import generateBitwiseMatrix from '../Routine/board.matrix.generateBitwise';
import getFlag from '../Routine/board.item.getFlag';
import getIsBoundary from '../Routine/board.item.getIsBoundary';
import getTotalItemsAround from '../Routine/board.item.getTotalItemsAround';
import getValue from '../Routine/board.item.getValue';
import isCoordinatesValid from '../Routine/board.coordinates.isValid';
import isValid from '../Routine/board.matrix.isValid';
import listBorderItems from '../Routine/region.board.listBorderItems';
import listEmptyItems from '../Routine/region.board.listEmptyItems';

export default class Board {
    private readonly data: number[][];

    public readonly height: number;
    public readonly width: number;

    private modelMines: boolean[][];
    private modelEmpty: boolean[][];

    public constructor(data: number[][], initialize: boolean = true)
    {
        if (! Board.isValid(data)) {
            throw Error();
        }

        this.data = data;
        this.height = data.length;
        this.width = data[0].length;

        if (initialize) {
            this.initialize();
            this.flagMines();
        }
    }

    public getValue(y: number, x: number): number
    {
        return getValue(this.data, y, x, this.height, this.width);
    }

    public getIsBoundary(y: number, x: number): boolean
    {
        return getIsBoundary(this.data, y, x, this.height, this.width);
    }

    public getIsMine(y: number, x: number): boolean
    {
        return getFlag(this.modelMines, y, x, this.height, this.width);
    }

    public setIsMine(y: number, x: number, flag: boolean = true): void
    {
        if (isCoordinatesValid(y, x, this.height, this.width)) {
            this.modelMines[y][x] = flag;
        }
    }

    public get countMines(): number
    {
        return countFlags(this.modelMines, this.height, this.width);
    }

    public getIsEmpty(y: number, x: number): boolean
    {
        return getFlag(this.modelEmpty, y, x, this.height, this.width);
    }

    public setIsEmpty(y: number, x: number, flag: boolean = true): void
    {
        if (isCoordinatesValid(y, x, this.height, this.width)) {
            this.modelEmpty[y][x] = flag;
        }
    }

    public get countEmpty(): number
    {
        return countFlags(this.modelEmpty, this.height, this.width);
    }

    public countMinesAround(y: number, x: number): number
    {
        return countFlagsAround(this.modelMines, y, x, this.height, this.width);
    }

    public countEmptyItemsAround(y: number, x: number): number
    {
        return countFlagsAround(this.modelEmpty, y, x, this.height, this.width);
    }

    public getTotalItemsAround(y: number, x: number): number
    {
        return getTotalItemsAround(y, x, this.height, this.width);
    }

    public get isConsistent(): boolean
    {
        for (let i: number = 0; i < this.height; i++) {
            for (let j: number = 0; j < this.width; j++) {
                if (this.getValue(i, j) < 0) {
                    continue;
                }

                if (this.getValue(i, j) < this.countMinesAround(i, j)) {
                    return false;
                }

                if (this.getTotalItemsAround(i, j) - this.countEmptyItemsAround(i, j) < this.getValue(i, j)) {
                    return false;
                }
            }
        }

        return true;
    }

    public get listEmptyItems(): Coordinate[]
    {
        return listEmptyItems(this);
    }

    public get listBorderItems(): Coordinate[]
    {
        return listBorderItems(this);
    }

    private initialize(): void
    {
        this.modelMines = generateBitwiseMatrix(this.height, this.width);
        this.modelEmpty = generateBitwiseMatrix(this.height, this.width);
    }

    private flagMines(): void
    {
        this.modelMines = flagMines(this.modelMines, this.data, this.height, this.width);
    }

    private static isValid(state: number[][]): boolean
    {
        return isValid(state);
    }
}
