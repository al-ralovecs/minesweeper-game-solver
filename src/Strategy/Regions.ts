import Board from '../Entity/Board';
import Coordinate from '../Entity/Coordinate';

export default class Regions {
    private readonly board: Board;
    private readonly bruteForceLimit: number;

    private data: Coordinate[][] = [];

    public constructor(board: Board, bruteForceLimit: number = 8)
    {
        this.board = board;
        this.bruteForceLimit = bruteForceLimit;
    }

    public get list(): Coordinate[][]
    {
        return this.data;
    }

    public get(n: number): Coordinate[]
    {
        if (0 <= n && this.count > n) {
            return this.data[n];
        }

        throw Error();
    }

    public get count(): number
    {
        return this.data.length;
    }

    public segregate(): void
    {

    }

    private doSegregate(tiles: Coordinate[]): Coordinate[][]
    {
        let allRegions: Coordinate[][] = [];
        let covered: Coordinate[] = [];

        while (true) {
            let queue: Coordinate[] = [];
            let finishedRegion: Coordinate[] = [];

            for (let t = 0; t < tiles.length; t++) {
                if (0 === covered.filter(e => e.value === tiles[t].value).length) {
                    queue.push(tiles[t]);
                    break;
                }
            }

            if (0 === queue.length) {
                break;
            }

            while (0 !== queue.length) {
                const currentTile: Coordinate = queue.pop();
                const ci: number = currentTile.getFirst;
                const cj: number = currentTile.getSecond;

                finishedRegion.push(currentTile);
                covered.push(currentTile);

                for (let t = 0; t < tiles.length; t++) {
                    const tile: Coordinate = tiles[t];
                    const ti: number = tile.getFirst;
                    const tj: number = tile.getSecond;

                    let isConnected: boolean = false;

                    if (0 !== finishedRegion.filter(e => e.value === tile.value).length) {
                        continue;
                    }

                    if (Math.abs(ci - ti) <= 2 || Math.abs(cj - tj) <= 2) {
                        tileSearch:
                        for (let i: number = 0; i < this.board.height; i++) {
                            for (let j: number = 0; j < this.board.width; j++) {
                                if (this.board.getValue(i, j) > 0) {
                                    if (Math.abs(ci - i) <= 1
                                        && Math.abs(cj - j) <= 1
                                        && Math.abs(ti -i) <= 1
                                        && Math.abs(tj - j) <= 1
                                    ) {
                                        isConnected = true;
                                        break tileSearch;
                                    }
                                }
                            }
                        }

                        if (! isConnected) {
                            continue;
                        }

                        if (0 === queue.filter(e => e.value === tile.value).length) {
                            queue.push(tile);
                        }
                    }
                }
            }

            allRegions.push(finishedRegion);
        }

        return allRegions;
    }

    private get getAllEmptyTiles() {
        let result: Coordinate[] = [];

        for (let i = 0; i < this.board.height; i++) {
            for (let j = 0; j < this.board.width; j++) {
                if (this.board.getValue(i, j) === -1 && ! this.board.isMine(i, j)) {
                    result.push(new Coordinate(i, j));
                }
            }
        }

        return result;
    }

    private get getBorderTiles() {

    }
}
