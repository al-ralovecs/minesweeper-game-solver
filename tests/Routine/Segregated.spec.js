import Board from '../../src/Entity/Board.ts';
import Regions from '../../src/Strategy/Regions';
import state from '../fixtures/state/11x16.json';

describe('Segregated', () => {
    test('check if segregates', () => {
        const board = new Board(state);
        const segregated = new Regions(board);
        segregated.segregate();

        expect(segregated.count).toBe(2);
        expect(segregated.list).toMatchSnapshot();
    })
});
