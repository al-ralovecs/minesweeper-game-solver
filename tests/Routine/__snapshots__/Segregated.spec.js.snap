// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Segregated check if segregates 1`] = `
Array [
  Array [
    Coordinate {
      "first": 0,
      "second": 6,
    },
    Coordinate {
      "first": 1,
      "second": 6,
    },
    Coordinate {
      "first": 3,
      "second": 6,
    },
    Coordinate {
      "first": 5,
      "second": 6,
    },
    Coordinate {
      "first": 7,
      "second": 7,
    },
    Coordinate {
      "first": 8,
      "second": 7,
    },
    Coordinate {
      "first": 6,
      "second": 7,
    },
    Coordinate {
      "first": 5,
      "second": 7,
    },
    Coordinate {
      "first": 4,
      "second": 6,
    },
    Coordinate {
      "first": 2,
      "second": 6,
    },
  ],
  Array [
    Coordinate {
      "first": 10,
      "second": 7,
    },
    Coordinate {
      "first": 12,
      "second": 7,
    },
    Coordinate {
      "first": 14,
      "second": 7,
    },
    Coordinate {
      "first": 14,
      "second": 6,
    },
    Coordinate {
      "first": 15,
      "second": 6,
    },
    Coordinate {
      "first": 13,
      "second": 7,
    },
    Coordinate {
      "first": 11,
      "second": 7,
    },
  ],
]
`;
